### PR TITLE
NEON implementation for Adler32

### PR DESCRIPTION
**Pull request by Adenilson**
The checksum is calculated in the uncompressed PNG data and can be made much faster by using SIMD.

Tests in ARMv8 yielded an improvement of about 3x (e.g. walltime was 350ms x 125ms for 4096x4096 bytes executed 30 times). That results in at least 18% improvement in PNG image decoding in Chromium.

Further details at:
https://bugs.chromium.org/p/chromium/issues/detail?id=688601

### CODE REVIEW

**Comment by ProgramMax** [comment]
Hello everyone. cblume@google.com / @chromium.org here.

Don't forget to add yourself to contrib/README.contrib

**Comment by Adenilson** [comment]
@ProgramMax nice catch! Fixed.

**Comment by Adenilson** [comment]
Just uploaded a function that uses the new ARMv8 crc32 instruction (it is about 10x faster than the C function in zlib).

To build using the optimized versions, just have an ARM compiler (e.g. from linaro, https://releases.linaro.org/components/toolchain/binaries/6.3-2017.02/) and export CC to point to it (e.g. export CC=arm-linux-gnueabihf-gcc-5) and next enable the feature in CMake buildsystem, either using the ncurses GUI (i.e. ccmake ..) or passing the options as:  
cmake .. -DARMv8=ON -DARMv8CRC=ON

I didn't touch the configure script and tried my best to integrate the ARM specific functions using the same strategy used for AMD64 and ASM686 specific code.

**Comment by Adenilson** [comment]
Also validated the changes with 'make test':
(xenial)adenilson@localhost:~/canonical-fork/build$ make test
Running tests...
Test project /home/adenilson/canonical-fork/build
    Start 1: example
1/2 Test #1: example ..........................   Passed    0.01 sec
    Start 2: example64
2/2 Test #2: example64 ........................   Passed    0.01 sec

100% tests passed, 0 tests failed out of 2

Total Test time (real) =   0.01 sec

**Comment by Adenilson** [comment]
Touching the base on this one as it has passed over 1 month already.

**Comment by Adenilson** [comment]
Friendly ping on the matter.

**Comment by Adenilson** [comment]
@madler anything you like to change on the original patch?

**Comment by Adenilson** [comment]
@madler ping?

**Comment by diizzyy** [comment]
@Adenilson 
Is this ARMv8 only or does it also work on ARMv7?

**Comment by Adenilson** [comment]
First patch should work in both ARMv7 (assuming that the SoC has a NEON unit) + ARMv8.

Second (CRC32) is ARMv8 (both AArch32 and AArch64) only.

**Comment by Adenilson** [comment]
@madler friendly ping?

**Comment by Adenilson** [comment]
For record, Chrome M62 is shipping the inflate_fast optimization and Chrome M63 has a variant of the Adler-32 optimization.

I'm working towards having the patch using the ARMv8 crc32 instruction included in M64 (branching in 2 weeks).

File: contrib/arm/armv8_crc32.c [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+// Depending on the compiler flavor, size_t may be defined in
+// one or the other header. See:
+// http://stackoverflow.com/questions/26410466/gcc-linaro-compiler-throws-error-unknown-type-name-size-t
+#include <stdint.h>
+#include <stddef.h>
+
+uint32_t armv8_crc32_little(uint32_t crc,
+                            const unsigned char *buf,
+                            size_t len) {
+    uint32_t c;
```
Comment (Adenilson) [inline contrib/arm/armv8_crc32.c]: In a second thought, this should handle the case of buf == Z_NULL.

**Comment by Adenilson** [comment]
An update on the issue: the crc32 optimization landed for Chrome M64 but it was reverted because it broke the build for android_x86 and android_x64.

I'm working in a new version for the crc32 optimization, using __crc32d() as it seems to be up to 2x faster than the original ARMv8 crc32 code (which was itself 10x faster than the vanilla C code).

I've posted some data in:
https://bugs.chromium.org/p/chromium/issues/detail?id=709716#c22

It can be from 32x to 45x faster (big core X little core) than the original zlib C function for vectors of 8KB.

As soon we land this on Chromium I'm planning to update this merge request.

Some samples:
--
random data length 16384 bytes
--test--  crc32     Min time     Max Rate    Median time  Median Rate
crc32_zlib d2a888f6     65 u-secs   0.24 MB/s     65 u-secs   0.24 MB/s
crc32_neon d2a888f6      1 u-secs  15.62 MB/s      1 u-secs  15.62 MB/s

random data length 8192 bytes
--test--  crc32     Min time     Max Rate    Median time  Median Rate
crc32_zlib 09bc142e     32 u-secs   0.24 MB/s     33 u-secs   0.24 MB/s
crc32_neon 09bc142e      0 u-secs    inf MB/s      1 u-secs   7.81 MB/s

**Comment by tbeu** [comment]
> Any hope to merge this upstream?

Hope never dies.

**Comment by Adenilson** [comment]
We did further testing, the **average** gain of the ARMv8 crc32 in PNG decoding ranges from 2.1% (140PNGs corpus) to 3.2% (Doodles) and up to 7.1% (Kodak).

The latest version of the crc32 patch (https://chromium-review.googlesource.com/c/chromium/src/+/801108) has code to perform the CPU feature detection on ARM.

**Comment by Adenilson** [comment]
The image corpus:
a) Kodak: http://r0k.us/graphics/kodak/
b) doodles: https://drive.google.com/drive/folders/1BaqNYg9jUbuUlYLxaBDLCo8him30qyZV?usp=sharing

**Comment by Adenilson** [comment]
The 140PNGs corpus is internal to Chromium developers (and access is granted by request).

**Comment by Adenilson** [comment]
@madler friendly ping on the subject.

The crc32 optimized function improved the performance decompressing gzipped content in 29%.

**Comment by Adenilson** [comment]
Hey, look at this... we are just about to complete 1 year (since the pull request)!
:-)

**Comment by Adenilson** [comment]
The ARMv8-a optimized crc32 code (an improved version of it) landed in Chromium almost 1 month ago and made the cut to Chromium M66.

If all goes well, M66 will be shipping to users in the first week of May with this optimization enabled for ARM.

**Comment by Adenilson** [comment]
Look at this, time really flies: 1 (one) year since the request was open!

To celebrate the anniversary, I opened another merge request (the inflate_fast NEON optimization, should be around 20% faster at decompression in average) at:
https://github.com/madler/zlib/pull/345

And if one day we manage to merge the Chromium optimizations (ARM + Intel):
https://github.com/madler/zlib/issues/346

**Comment by Adenilson** [comment]
Friendly ping...

**Comment by Adenilson** [comment]
@madler any comment?

**Comment by Adenilson** [comment]
It is being over 1 1/2 year (i.e. 18 months), I'm closing this merge request for the given reasons:

a) We are shipping faster (and better tested) versions of this optimizations on Chromium.
b) These optimizations have also being ported to x86.

For anyone interested, please check the latest code in:
a) NEON-ized Adler-32: https://cs.chromium.org/chromium/src/third_party/zlib/adler32_simd.c?l=203
b) ARMv8 crypto accelerated CRC32: https://cs.chromium.org/chromium/src/third_party/zlib/crc32_simd.c?l=159

**Comment by tbeu** [comment]
Too bad.

**Comment by praiskup** [comment]
Mark, I'm curious what would help with reviews of similar contributions like this one:
- is this something which should never go to zlib?
- is this lack of time only? Could you e.g. specify how to become a contributor to zlib, so you could enlarge the set of people you trust enough to do the code reviews?
- would some automatic testing help here?  I mean, if we had some HW being able to test the added code, would you be more confident that the code from pull request is acceptable?

**Pull request closed without merging**